Feature/#3 Loudy API 통합을 위한 네트워크 모듈 구현

simuelunbo opened the pull request:
#3 
두서 없이 개발을 진행 하다보니 network 모듈과 함꼐 data 모듈 일부도 추가하였습니다.
그래서 코드 리뷰 내용양이 너무 많을 것 같은데 다음번에는 확실하게 나눠서 분리하여 PR 올리도록 하겠습니다 🙇

추가적으로 궁금한게 있는데 현재 network 모듈에 model에 
```
This declaration is opt-in and its usage should be marked with @kotlinx.serialization.InternalSerializationApi or @OptIn(markerClass = kotlinx.serialization.InternalSerializationApi.class
```
해당 warning이 나오고 있습니다.
그래서 20ce24053c3c42724413c4eed3be2601a6a4bc38에 보면 configureKotlin 확장 함수에  "-opt-in=kotlinx.serialization.InternalSerializationApi" 추가 하였는데도 사라지지 않고 있습니다 무엇이 문제인걸까요? 
리뷰 잘 부탁드립니다.

simuelunbo committed:
build: 프로젝트 설정 파일 업데이트

- gitignore: http-client.private.env.json 파일 제외 추가
- build-logic: Kotlin 컴파일러 옵션 최적화
- gradle: 의존성 라이브러리 버전 업데이트
- settings: core 모듈 포함 설정 추가

simuelunbo committed:
refactor: 불필요한 API 테스트 파일 삭제

simuelunbo committed:
feat: core:data 모듈 기본 구조 생성

- 데이터 레이어 모듈 기본 구조 생성
- build.gradle.kts 의존성 설정
- AndroidManifest.xml 설정

simuelunbo committed:
feat: core:data 도메인 모델 클래스 구현

- AccountLimitInfoDto: 계정 한도 정보 모델
- SongDto: 음악 정보 모델 (MusicKeyDto 포함)
- SongGenerationDto: 음악 생성 요청 모델
- RandomPromptDataDto: 랜덤 프롬프트 모델

simuelunbo committed:
feat: core:data 데이터소스 인터페이스 구현

- LoudlyRemoteDataSource: 원격 데이터소스 인터페이스
- 계정 한도 조회, 음악 생성, 랜덤 프롬프트 조회 메서드 정의
- suspend 함수로 코루틴 지원

simuelunbo committed:
feat: core:network 모듈 기본 구조 생성

- 네트워크 레이어 모듈 기본 구조 생성
- build.gradle.kts 의존성 설정 (Retrofit, OkHttp, Hilt 등)
- 네트워크 라이브러리 의존성 추가

simuelunbo committed:
feat: core:network 모델 및 예외 처리 클래스 구현

- 네트워크 응답 모델 클래스 구현
- AccountLimitResponse, SongResponse, RandomPromptResponse 모델
- GenerateSongRequest, ErrorResponse 모델
- ApiException 커스텀 예외 클래스 구현

simuelunbo committed:
feat: core:network API 인터페이스 및 인터셉터 구현

- LoudlyApiService: Retrofit API 인터페이스 정의
- ApiKeyInterceptor: API 키 헤더 추가 인터셉터
- ErrorHandlingInterceptor: 에러 응답 처리 인터셉터
- 음악 생성, 계정 한도, 랜덤 프롬프트 API 엔드포인트 정의

simuelunbo committed:
feat: core:network 데이터소스 구현체 구현

- LoudlyRemoteDataSourceImpl: 원격 데이터소스 구현체
- Retrofit API 서비스를 통한 네트워크 호출
- 네트워크 응답을 도메인 모델로 변환
- 날짜 파싱 및 데이터 매핑 로직 구현

simuelunbo committed:
feat: core:network 의존성 주입 모듈 구현

- NetworkModule: Retrofit, OkHttp, 인터셉터 의존성 제공
- DataSourceModule: 데이터소스 구현체 바인딩
- Hilt 모듈로 의존성 주입 설정
- API 키 및 네트워크 설정 구성

simuelunbo committed:
security: http-client.env.json 파일 gitignore 추가

- API 키가 포함된 환경 설정 파일 제외
- 보안상 중요한 파일이 커밋되지 않도록 방지

simuelunbo committed:
test: core 모듈 테스트 코드 구현

- FakeLoudlyApiService: 테스트용 API 서비스 구현
- LoudlyRemoteDataSourceImplTest: 데이터소스 단위 테스트
- NetworkIntegrationTest: 네트워크 통합 테스트
- 예외 처리 및 성공 케이스 테스트 포함
- runBlocking을 사용한 코루틴 테스트 구현

simuelunbo committed:
docs: HTTP 클라이언트 테스트 파일 추가

- MyApi.http: API 엔드포인트 테스트용 HTTP 파일
- 개발 및 테스트 목적의 API 호출 샘플

f-lab-loco commented on core/network/src/main/java/com/simuel/musicgenerator/core/network/exception/ApiException.kt:
```diff
@@ -0,0 +1,8 @@
+    val code: Int,
+    val error: String,
+    val errorDescription: String,
+    val payload: String? = null
```
payload는 어떤 내용을 생각하고 계시는걸까요?

simuelunbo replied:
https://www.loudly.com/developers/paths/api-ai-prompt-songs/post
api 문서에서 500으로 Error 응답시 payload 필드가 정의 되어 있어 사용하고 있습니다.

f-lab-loco commented on core/network/src/test/java/com/simuel/musicgenerator/core/network/api/FakeLoudlyApiService.kt:
```diff
@@ -0,0 +1,76 @@
+import com.simuel.musicgenerator.core.network.model.RandomPromptResponse
+import com.simuel.musicgenerator.core.network.model.SongResponse
+
+internal class FakeLoudlyApiService : LoudlyApiService {
```
테스트용으로 만들어야 하나요? 단순히 mocking만으로 처리한것과의 차별점이 있나요?

simuelunbo replied:
네 mocking과 fake용으로 만드는것은 차이가 있습니다. 
mock은 '해당 기능을 호출 하였는가?'를 검증 하는것이고 fake는 '해당 기능의 결과과 expect와 같은가?'를 검증하는 것입니다. 
굳이 fake 객체를 만들지 않고 mock으로 처리해도 되지만 fake 구현체로도 한번 만들어보고 싶어서 진행을 해보았습니다.

f-lab-loco commented on core/network/src/main/java/com/simuel/musicgenerator/core/network/di/NetworkModule.kt:
```diff
@@ -0,0 +1,73 @@
+    }
+    
+    @Provides
+    @Singleton
```
어차피 NetworkModule이 object라서 sington annotation이 필요 없을것 같아요.
설정하신것과 설정제거한것의 generate 된 코드를 보면 어떤차이가 있을까요?

simuelunbo replied:
8a3a7ae7d8999e73e3b301a8c30c34945dda5cea 수정하였습니다.

generate 된 코드를 확인을 해보니 다음과 같았습니다.
싱글톤 어노테이션을 제거 했을때와 제거 하지 않았을때 차이는 @ScopeMetadata 어노테이션 입니다.
싱글톤 어노테이션이 있을 경우 @ScopeMetadata("javax.inject.Singleton")
싱글톤 어노테이션이 없을 경우 @ScopeMetadata 
다른 부분은 동일하게 나오고 있습니다.
이 차이는 dagger에서 생성한 factory 클래스가 어떤 스코프에서 동작할지 나타내는 것입니다.
싱글톤 어노테이션이 있으면 싱글톤으로 관리가 되고 반복 되는 호출 같은 경우엔 캐시된 인스턴스를 반환합니다., 그렇지 않은 경우 호출될 떄마다 새로운 인스턴스를 생성 합니다.

f-lab-loco reviewed (commented):
나중에 network 와 module 분리되면 network 관련 코드에서 internal 이 몇개 빠질 것 같습니다. 참고 해주세요

f-lab-loco commented:
```
추가적으로 궁금한게 있는데 현재 network 모듈에 model에

This declaration is opt-in and its usage should be marked with @kotlinx.serialization.InternalSerializationApi or @OptIn(markerClass = kotlinx.serialization.InternalSerializationApi.class
해당 warning이 나오고 있습니다.
그래서 20ce24053c3c42724413c4eed3be2601a6a4bc38에 보면 configureKotlin 확장 함수에 "-opt-in=kotlinx.serialization.InternalSerializationApi" 추가 하였는데도 사라지지 않고 있습니다 무엇이 문제인걸까요?
리뷰 잘 부탁드립니다.
```

https://github.com/Kotlin/kotlinx.serialization/issues/2844#issuecomment-2443896218
이거 문서 한번 보시겠습니까?

simuelunbo committed:
refactor: NetworkModule 싱글톤 어노테이션 제거

This change is made because Dagger already manages the lifecycle of
dependencies provided by the `NetworkModule`, and explicitly marking
this function as a singleton is redundant.

f-lab-loco commented on core/network/src/main/java/com/simuel/musicgenerator/core/network/di/NetworkModule.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    @Provides
+    internal fun provideLoudlyApiService(retrofit: Retrofit): LoudlyApiService {
```
`fun provideLoudlyApiService()` 으로 만들어도 될 것 같은데 `fun provideLoudlyApiService(retrofit: Retrofit)` 파라미터를 추가해서 Retrofit을 관리하신 이유가 있으신거죠?

simuelunbo replied:
@f-lab-loco 좋은 지적 감사합니다. 다른 프로젝트 들에서 일반적으로 사용하는 패턴을 따라서 관성처럼 코드를 작성 한 것이였습니다. 
현재 프로젝트 구조에서는 굳이 파라미터를 받을 필요가 없겠네요 
문제가 있으면 추후 코드 리팩토링 하도록 하겠습니다.

f-lab-loco replied:
음.. 변경을 하셔도되긴하는데, 코드 커버리지측면은 현상태로 유지하시는게 조금 더 테스트가 가능해져서 코드 커버리지가 올라갈꺼에요. 전 의도 하신줄 알았어요 ㅎㅎ
